packages/hardhat/contracts/periphery/EulerControllerHelper.sol

```diff
@@ -0,0 +1,103 @@
+// SPDX-License-Identifier: agpl-3.0
+pragma solidity =0.7.6;
+pragma abicoder v2;
+
+import "hardhat/console.sol";
+
+// interface
+import {IERC20Detailed} from "../interfaces/IERC20Detailed.sol";
+
+// lib
+import {SafeMath} from "@openzeppelin/contracts/math/SafeMath.sol";
+
+/// @notice Definition of callback method that deferLiquidityCheck will invoke on your contract
+interface IDeferredLiquidityCheck {
+    function onDeferredLiquidityCheck(bytes memory data) external;
+}
+
+interface IExec {
+    function deferLiquidityCheck(address account, bytes memory data) external;
+}
+
+interface IEulerDToken {
+    function borrow(uint256 subAccountId, uint256 amount) external;
+
+    function repay(uint256 subAccountId, uint256 amount) external;
+}
+
+contract EulerControllerHelper is IDeferredLiquidityCheck {
+    using SafeMath for uint256;
+
+    address public immutable exec;
+    address public immutable euler;
+    address public immutable token;
+    address public immutable dToken;
+
+    struct FlashloanCallbackData {
+        address caller;
+        address assetToBorrow;
+        uint256 amountToBorrow;
+        uint8 callSource;
+        bytes callData;
+    }
+
+    constructor(
+        address _exec,
+        address _euler,
+        address _token,
+        address _dToken
+    ) {
+        exec = _exec;
+        euler = _euler;
+        token = _token;
+        dToken = _dToken;
+    }
+
+    function _flashCallback(
+        address _initiator,
+        address _asset,
+        uint256 _amount,
+        uint256 _premium,
+        uint8 _callSource,
+        bytes memory _calldata
+    ) internal virtual {}
+
+    function onDeferredLiquidityCheck(bytes memory encodedData) external override {
+        // sanity checks
+        require(msg.sender == euler);
+
+        console.log("msg.sender", msg.sender);
+
+        FlashloanCallbackData memory data = abi.decode(encodedData, (FlashloanCallbackData));
+
+        IEulerDToken(dToken).borrow(0, data.amountToBorrow);
+
+        _flashCallback(data.caller, data.assetToBorrow, data.amountToBorrow, 0, data.callSource, data.callData);
```

Comment:
Imo should kill the _premium param here and in ControllerHelper to save gas.

[#337](https://github.com/opynfinance/squeeth-monorepo/issues/337)